### DISCUSSION TITLE
Proper way to do machine-to-machine authentication?

### DISCUSSION THREAD

**Discussion by davidsickmiller** [symfony/symfony#52151] [Q&A]
I'm using Symfony for a B2B SaaS use case, and I'm curious how I should be doing API key authentication.

I've followed the [ApiKeyAuthenticator sample code](https://symfony.com/doc/current/security/custom_authenticator.html), and I've given API keys to my users, who have embedded them in their API clients.

The problem happens after a year, when Bob The Programmer at Client Corporation quits his job.  Client Corporation duly deactivates his user account, and because the API is associated with his user account, their API calls stop working.

For a small, new SaaS company, this problem happens rarely enough.  But it becomes a problem at scale and over time.

I'm curious if there's a "Symfony Way" of dealing with the problem.

One option would be to tell all the client companies that they should rarely use API keys associated with human user accounts but instead create non-human user accounts for API keys that need to persist through employee job changes.  This is not an unheard-of practice; I believe this how Salesforce does it.

A variation of the above option would be to do it, but hide it from the client -- automatically create a user record for machine API keys, exclude them from the user management UI to avoid confusion, etc.

Another approach would be to more formally recognize identities that have permissions but aren't associated with specific people.  (An example of this is [IAM roles](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles.html) in AWS.).  I haven't found this concept in the Symfony Docs.  I found [one StackOverflow answer showing a way to do it](https://stackoverflow.com/a/75538413/718475), and I assume it works, but it feels hacky.

I've been searching about this for a while but haven't found a lot.  Maybe I'm using the wrong words to describe it?

**Comment by ghost** [comment]
There is no "Symfony way". This is not a question that a framework answers. It's the businness/technical/security requirements that lead to the decision what type of authentication is chosen.

> One option would be to tell all the client companies that they should rarely use API keys associated with human user accounts but instead create non-human user accounts for API keys that need to persist through employee job changes. 

Yes, this is a valid solution. If your client (the company) creates an account let's say by the developer that codes the integration with your app it means they are not very mature as a company. They should create some general account and credentials should be written down on their internal Wiki (e.g. Confluence) or (better and safer) the account should be created by company's administrator or some other person of trust or non-rank and file employee.

If I understood the description correctly, to get an API key there must be an account created on your website. If so and it is a root of the whole problem then just don't require creating an account. Create API Key manually, add to your database and then give it to your client. That's it. This still has a drawback that ex-empolyees of your client can use your API after the leave so at least rotate this API Key every x time. For better solution you could introduce IP address authentication, i.e. client gives you the IP address that he will call your API from (I won't elaborate about solving organizational problem like prod vs dev env where IPs are obviously different [dev ip might be even dynamic] - that could be solved by company's VPN and developer would have to be connected to the VPN when calling your API)

**Reply by davidsickmiller** [reply]
@javaDeveloperKid , I appreciate the thoughtful reply, but I think you missed the point of my question.

In case this clarifies:

Ideally, I'd like to get the benefits of Symfony's security infrastructure, such as `@Security("is_granted('ROLE_XX')")` and the Firewall, for API keys that aren't tied to specific people.  Theoretically that should be possible, but it doesn't seem well supported by the framework -- the [Symfony Security page](https://symfony.com/doc/current/security.html) is quite long with detailed information about how the framework supports lots of scenarios, but doesn't seem to mention this one.  And it doesn't look like the Passport class was designed for cases where there isn't a user.

There's surely a way to do anything, but working with the framework, rather than against it, usually makes things simpler today as well as in the future when it comes time to upgrade to the next version.

**Reply by ghost** [reply]
> for API keys that aren't tied to specific people

Actually this is one of the main reasons why Security Component was rewritten im version 5. A concept of passports was introduced to decouple security from the user context. You still have UserInterface but it's more abstract than before i.e. it isn't a user in the meaning of physical person but **anything** that authenticates in your system. That's why is has only 3 methods `getRoles()`, `eraseCredentials()` (this one can be empty), `getUserIdentifier()`. You can create a concrete implementation of UserInterface for your authenticator.
Maybe this two articles will give you more view: 
https://wouterj.nl/2020/04/authenticators-new-symfony-security 
https://wouterj.nl/2019/03/security-removing-user

**Comment by yblatti** [comment]
> Actually this is one is the main reasons why Security Component was rewritten im version 5. A concept of passports was introduced to decouple security from the user context. You still have UserInterface but it's more abstract than before i.e. it isn't a user in the meaning of physical person but anything that authenticates in your system. That's why is has only 3 methods getRoles(), eraseCredentials() (this one can be empty), getUserIdentifier(). You can create a concrete implementation of UserInterface for your authenticator.

Can confirm. I do Exactly that for the remote systems that I want to authenticate 'like a User' but I don't want to 'be a User' and it works like a charm ! (Disclaimer: I'm not a Symfony expert, maybe someone will correct me :) )

**Main steps :** 

- Add a class for your 'Machine Users' : `class MachineUser implements UserInterface`
- Implement at least `getRoles()`and `getUserIdentifier()` on `MachineUser `
- Add a user provider : 
   ```yaml security:
    providers:
        app_user_provider: # your current users
            entity:
                class: App\Entity\User
        app_machine_provider: # new 'machine' users
            entity:
                class: App\Entity\MachineUser
                property: maybeASpecialUsername```

- Add a firewall for your API to use this provider, and your new ApiKeyAuthenticator : 
   ```yaml security:
    firewalls:
        m2m_api:
            custom_authenticators:
                - App\Security\ApiKeyAuthenticator # https://symfony.com/doc/current/security/custom_authenticator.html
            # or maybe #
            access_token:
                token_handler: App\Security\AccessTokenHandler # https://symfony.com/doc/current/security/access_token.html
            pattern: ^/api/m2m
            provider: app_machine_provider
            stateless: true
        main:
            provider: app_user_provider
            [...]```
- Boom, once you have MachineUsers with Keys or Token in your db and the needed ApiKeyAuthenticator or AccessTokenHandler, you'll be able to authenticate and to stuff like
   ```php 
      #[IsGranted('ROLE_M2M_API_READ')]
      #[Route(path: '/api/m2m/cryptoLosses']
      class CryptoLossesApiController extends AbstractController```
